File: Binary_Search/2602.Minimum_Operations_to_Make_All_Array_Elements_Equal/C++/2602.Minimum_Operations_to_Make_All_Array_Elements_Equal.cpp
class Solution {
public:
    vector<long long> minOperations(vector<int> & nums, vector<int> & queries) 
    {
        //=======================================//
        // Input type:                           //
        //   - vector<int> ref. (nums vector)    //
        //   - vector<int> ref. (queries vector) //
        // Return type:                          //
        //   - vector<long long> (result vector) //
        //=======================================//

        /*Initialize*/
        ///// Length of vectors (i.e. nums queries) /////
        int len_nums = nums.size(), len_queries = queries.size();

        ///// Sorted operations /////
        sort(nums.begin(), nums.end());

        ///// Record dynamic-programmig (i.e. DP) vector /////
        vector<long long> record_dp((len_nums + 1), 0);

        ///// Record index /////
        int record_idx;

        ///// Result vector /////
        vector<long long> res_vect(len_queries, 0);


        /*Binary-search based loop traversal with recorded dynamic-programming (i.e. DP) vector*/
        ///// Step 1: Record prefix-summary value with dynamic-programming vector /////
        for (int record_idx = 1; (record_idx < record_dp.size()); (record_idx++)) //Whole
        {
            record_dp[record_idx] = (record_dp[(record_idx - 1)] + nums[(record_idx - 1)]); //Keep updating, overwriting
             
        } //Whole

        ///// Step 2: Looped-traversal with recorded dynamic-programming vector /////
        for (int queries_idx = 0; (queries_idx < len_queries); (queries_idx++)) //Whole
        {
            record_idx = binarySearch(nums, len_nums, queries[queries_idx]); //Binary-search function call

            res_vect[queries_idx] = ((((long long)(queries[queries_idx]) * record_idx) - record_dp[record_idx]) + ((record_dp.back() - record_dp[record_idx]) - ((long long)(queries[queries_idx]) * (len_nums - record_idx)))); //Keep updating, overwriting

        } //Whole

        return res_vect;
    }


    int binarySearch(vector<int> & nums, int & len_nums, int & record_val)
    {
        //======================================//
        // Input type:                          //
        //   - vector<int> ref. (nums vector)   //
        //   - int ref. (length of nums vector) //
        //   - int ref. (record value)          //
        // Return type:                         //
        //   - int (result index)               //
        //======================================//

        /*Initialize*/
        ///// Record pointers (i.e. middle, left, right) /////
        int record_mid_ptr, record_left_ptr = 0, record_right_ptr = (len_nums - 1);

        /*Whole process, flow*/
        while (record_left_ptr <= record_right_ptr) //Binary-search
        {
            record_mid_ptr = ((record_left_ptr + record_right_ptr) / 2); //Keep updating, overwriting

            ///// Check if the current indexed-value matched conditions or not /////
            if (nums[record_mid_ptr] < record_val) { record_left_ptr = (record_mid_ptr + 1); } //Keep updating, overwriting

            else if (nums[record_mid_ptr] > record_val) { record_right_ptr = (record_mid_ptr - 1); } //Keep updating, overwriting
            else { return record_mid_ptr; }

        } //Binary-search

        return record_left_ptr;
    }
};

File: Binary_Search/2602.Minimum_Operations_to_Make_All_Array_Elements_Equal/Python/2602.Minimum_Operations_to_Make_All_Array_Elements_Equal.py
class Solution(object):
    def minOperations(self, nums, queries):
        """
        :type nums: List[int]
        :type queries: List[int]
        :rtype: List[int]
        """
        #======================================#
        # Binary-search based traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of arrays (i.e. nums, queries) #####
        len_nums, len_queries = len(nums), len(queries)

        ##### Sorted operations #####
        nums.sort()

        ##### Record dynamic-programming (i.e. DP) array #####
        record_dp = [0 for _ in range(len_nums + 1)]

        ##### Result array #####
        res_arry = [0 for _ in range(len_queries)]


        ######################################################################################
        #Binary-search based loop traversal with recorded dynamic-programming (i.e. DP) array
        ##### Step 1: Record prefix-summary value with dynamic-programming array #####
        for record_idx in range(1, (len_nums + 1)):
            record_dp[record_idx] = (record_dp[(record_idx - 1)] + nums[(record_idx - 1)]) #Keep updating/overwriting

        ##### Step 2: Looped-traversal with recorded dynamic-programming array #####
        for queries_idx in range(len_queries):
            record_idx = self.binarySearch(nums, len_nums, queries[queries_idx]) #Record index

            res_arry[queries_idx] = (((queries[queries_idx] * record_idx) - record_dp[record_idx]) + ((record_dp[(-1)] - record_dp[record_idx]) - (queries[queries_idx] * (len_nums - record_idx)))) #Keep updating/overwriting

        return res_arry


    def binarySearch(self, nums, len_nums, record_val):
        """
        :type nums: List[int]
        :type len_nums: int
        :type record_val: int
        :rtype: int
        """
        #========================================#
        # Binary-search based traversal function #
        #========================================#

        ############
        #Initialize
        ##### Record pointers (i.e. left, right) #####
        record_left_ptr, record_right_ptr = 0, (len_nums - 1)

        ####################
        #Whole process/flow
        while (record_left_ptr <= record_right_ptr):
            record_mid_ptr = ((record_left_ptr + record_right_ptr) // 2) #Record middle pointer

            ##### Check if the current indexed-value matched conditions or not #####
            if (nums[record_mid_ptr] < record_val):
                record_left_ptr = (record_mid_ptr + 1) #Keep updating/overwriting

            elif (nums[record_mid_ptr] > record_val):
                record_right_ptr = (record_mid_ptr - 1) #Keep updating/overwriting
            else:
                return record_mid_ptr

        return record_left_ptr
